README.md
user-script
=====

![Travis Status](https://travis-ci.org/ZCloud-Firstserver/user-script.png?branch=master)

See pro tips. [Continuous convergence with the chef-solo on the joyent smartmachine and meta-data API.](https://coderwall.com/p/9gk3ag)

Overview
----


SmartMachineをプロビジョン時にChef-SoloのCronを仕込んで継続的に実行する、一回限りのオプションにも対応。


Metadata
----

利用可能メタデータ一覧

- zcloud_app: アプリケーション名、**必須**
- zcloud_app_repo: アプリケーションリポジトリのGit、**必須**
- zcloud_app_ref: リモートのbranch名、省略時は`master`
- zcloud_app_once: 初回のみChefSoloを実行、2度目以降は何もしない。キーがあれば値はなんでも良い。
- user-data: Chefのノード用Json、必要ない場合は文字列のブレス `{}`、**必須**

- zcloud_hostname: hostnameに使用出来る文字、省略時はIPアドレス＋zonename
- zcloud_timezone： `sm-list-timezones` で取得できるもの、デフォルトはJapan
- zcloud_notify_to： カンマ区切りメールアドレス

Chef−Solo
----

#### Chef-repo

`Meadatta: zcloud_app_repo` にセットされたgitリポジトリをクローンし、以降Masterが継続的にフェッチされる。

#### run_list

`Meadatta: zcloud_app`にセットされたロール名でChef-Soloが実行される `role[${zcloud_app}]`

#### override_attributes(node.json)

`Meadatta: user-data`にOverride用のJsonを格納するとChef-soloが取り込んでくれる(`-j` オプション)。


Status
----

スクリプト実行段階に応じて`${CURRENTSTATE}`が変化する、レポートメールに記載される。

- initalize
- setup_host
- setup_wrapper
- setup_cronjob
- setup_ohai_plugin
- setup_chef-solo
- initalize_git_repository
- update_git_repository
- execute_chef-solo
- failure_chef-solo
- running

Report
----

以下の場合メールが送信される。

- zcloud_notify_to がセットされている
- chef-solo初回起動時のみ結果を通知
- user-script実行時、途中でエラー終了した場合(初回以降は約10分おき実行)

Aplication Examples
----

User-Scriptと連携するサンプルリポジトリを用意しました。

- **Dokuwiki** : `[https://github.com/ZCloud-Firstserver/application_dokuwiki](https://github.com/ZCloud-Firstserver/application_dokuwiki)

Contributing
------------

e.g.

1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write you change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
Authors: sawanoboryu@higanworks.com (HiganWorks LLC)

Licensed under the Apache License, Version 2.0 (the "License");


CHANGELOG.md
# CHANGELOG of the user-script


## v2.0

- バージョン番号が付きました
- ChangeLogをつけ始めました


## - v1.1

- 正式版

test/travis.sh
#!/usr/bin/env sh
bash -n *.sh && echo Syntax OK


base64.sh
#!/usr/bin/bash

set -e
set -x

## set specific PATH
if [ -f /etc/product ]
then
  SM_VERS=`grep "^Image" /etc/product | awk '{print $3}'`
  case "$SM_VERS" in
    1.8.?)
      export PATH=/opt/local/gnu/bin:/opt/local/bin:/opt/local/sbin:/usr/bin:/usr/sbin
      ;;
    *)
      export PATH=/opt/local/bin:/opt/local/sbin:/usr/bin:/usr/sbin
      ;;
  esac
fi

## Current state of povisioning
CURRENTSTATE=initalize
COMPLATEFILE=/var/svc/zcloud_success
MAILFLAG=true
T_MAILBODY="/tmp/$(basename $0).$$.tmp"
touch $T_MAILBODY


_finalize() {
  EXITCODE=$?
  cat <<EOL >> $T_MAILBODY
Hi,

This is zcloud application automator.

My ipaddress is $IPADDRESS.
Last state is $CURRENTSTATE and exit status is $EXITCODE.

EOL
  echo >> $T_MAILBODY
  echo "========== tail of metadata-execute log" >> $T_MAILBODY
  tail /var/svc/log/smartdc-mdata\:execute.log >> $T_MAILBODY
  if $MAILFLAG
  then
    cat $T_MAILBODY | mailx -s "[$CURRENTSTATE $Z_APP] Zcloud Notification from `hostname`" $MAILTO
  fi
  rm $T_MAILBODY
  exit 0
}

trap _finalize 0

## shared functions
_smf_enabler() {
  if svcs "$1" | grep -q disabled  ; then svcadm enable "$1" ;fi
  if svcs "$1" | grep -q maintenance  ; then svcadm clear "$1" ;fi
}

_get_addr_by_if() {
  ipadm show-addr "$1" -p -o ADDR
}

_mdata_check(){
  if ! mdata-get $1 ; then echo "ERROR_EXIT: missing metadata $1" ; exit 1 ; fi
  export $2="`mdata-get $1`"
}

_splitip() {
  local IFS="/"
  set -- $1
  echo $1
}

# postfix must be running.
_smf_enabler postfix

MAILTO=""
## Set localzone and force reboot
if mdata-get zcloud_notify_to
then
  MAILTO=`mdata-get zcloud_notify_to`
else
  MAILFLAG=false
fi

###
### prepare section
###

CURRENTSTATE=setup_host

IPADDRESS=`_get_addr_by_if net0/_a`

## set hostname. It is no need to reboot.
if mdata-get zcloud_hostname
then
  LNAME=`mdata-get zcloud_hostname`
  if [ ! "`hostname`" == "$LNAME" ] ; then sm-set-hostname ${LNAME} ; fi
else
  ZONENAME=`zonename`
  LNAME=`_splitip ${IPADDRESS//./-}`.${ZONENAME:0:8}.local
  if [ ! "`hostname`" == "$LNAME" ] ; then sm-set-hostname ${LNAME} ; fi
fi

## Set localzone and force reboot
if mdata-get zcloud_timezone
then
  LZONE=`mdata-get zcloud_timezone`
else
  LZONE=Japan
fi
if [ ! "$TZ" == "$LZONE" ] ; then MAILFLAG=false ; sm-set-timezone ${LZONE} && reboot ; fi

###
### main section
###

MDATA_WRAPPER=001
MDATA_USERSCRIPT=/var/svc/mdata-user-script
MDATA_USERDATA=/var/svc/mdata-user-data
CHEF_REPOS=/usr/local/zcloud-application

## add mdata-wapper script to cron
## Notice: this block must keep on top to retry fetch cyclically.


if [ ! -f /opt/local/sbin/mdata_wrapper_${MDATA_WRAPPER}.sh ] ; then
  CURRENTSTATE=setup_wrapper
  cat <<"EOL" > /opt/local/sbin/mdata_wrapper_${MDATA_WRAPPER}.sh
#!/usr/bin/bash
export PATH=/opt/local/bin:/opt/local/sbin:/usr/bin:/usr/sbin
sleep `expr $$ % 60`
svcadm restart mdata:fetch
if svcs mdata:fetch | grep -q maintenance  ; then svcadm clear mdata:fetch ;fi
sleep 2
svcadm restart mdata:execute
if svcs mdata:execute | grep -q maintenance  ; then svcadm clear mdata:execute ;fi
EOL

  chmod u+x /opt/local/sbin/mdata_wrapper_${MDATA_WRAPPER}.sh
  ln -fs /opt/local/sbin/mdata_wrapper_${MDATA_WRAPPER}.sh /opt/local/sbin/mdata_wrapper.sh
fi

if ! grep -q -x "## ZCloud-Application" /var/spool/cron/crontabs/root
then
  CURRENTSTATE=setup_cronjob
  cat << "EOL" >> /var/spool/cron/crontabs/root
## ZCloud-Application
0,10,20,30,40,50 * * * * /opt/local/sbin/mdata_wrapper.sh
EOL
svcadm restart cron
fi

## notifies start setup if new server

if [ ! -f $COMPLATEFILE -a -z $MAILTO ]
then
  cat <<EOL | mailx -s "[Setup started]Zcloud Norify from `hostname`" $MAILTO
Hi,

This is zcloud application automator.

My ipaddress is $IPADDRESS.
A new server has been provisioning. Please wait a few minutes.

EOL
fi

# install joyent_attr_plugin
if [ ! -f /opt/local/etc/ohai/plugins/joyent.rb ] ; then
  CURRENTSTATE=setup_ohai_plugin
  install -d /opt/local/etc/ohai/plugins -m 0755
  curl -skf -o /opt/local/etc/ohai/plugins/joyent.rb https://raw.github.com/ZCloud-Firstserver/ohai_plugin_joyent/master/plugins/joyent.rb
fi

## install chef-solo
if [ ! -f /opt/local/bin/chef-solo ] ; then
  CURRENTSTATE=setup_chef-solo

  pkgin -y install gcc47 scmgit-base gmake ruby193-base ruby193-yajl ruby193-nokogiri ruby193-readline pkg-config

## for smf cookbook
  pkgin -y install libxslt

## install chef
  gem update --system
  gem install --no-ri --no-rdoc bundler
  gem install --no-ri --no-rdoc ohai
  gem install --no-ri --no-rdoc json
  gem install --no-ri --no-rdoc chef
  gem install --no-ri --no-rdoc rb-readline
fi

## get attribute from metadata-api

_mdata_check zcloud_app Z_APP
_mdata_check zcloud_app_repo Z_APP_REPO
if mdata-get zcloud_app_ref ; then _mdata_check zcloud_app_ref Z_APP_REF
else
    Z_APP_REF=master
fi

## exit once
if [ -f $COMPLATEFILE ]
then
  if mdata-get zcloud_app_once
  then
    CURRENTSTATE=running
    exit 0
  fi
fi


## clone or pull application repositoly to local

if [ ! -d ${CHEF_REPOS} ] ; then
  CURRENTSTATE=initalize_git_repository
  git clone ${Z_APP_REPO} ${CHEF_REPOS} -b ${Z_APP_REF}
else
  CURRENTSTATE=update_git_repository
  cd ${CHEF_REPOS}
  ## switch branch by Metadata
  git fetch --all --prune
  git checkout ${Z_APP_REF}
  git pull origin ${Z_APP_REF}
fi

## execute chef-solo
CURRENTSTATE=execute_chef-solo

if chef-solo -j ${MDATA_USERDATA} -c ${CHEF_REPOS}/solo.rb -o "role[${Z_APP}]"
then
  CURRENTSTATE=running
  if [ -f $COMPLATEFILE ]
  then
    MAILFLAG=false
  else
    touch $COMPLATEFILE
  fi
else
  CURRENTSTATE=failure_chef-solo
fi


exit 0


smartos19.sh
#!/usr/bin/bash
## version: v2.0
## support: https://github.com/ZCloud-Firstserver/user-script

set -e
set -x

## set specific PATH
if [ -f /etc/product ]
then
  SM_VERS=`grep "^Image" /etc/product | awk '{print $3}'`
  case "$SM_VERS" in
    1.8.?)
      export PATH=/opt/local/gnu/bin:/opt/local/bin:/opt/local/sbin:/usr/bin:/usr/sbin
      ;;
    *)
      export PATH=/opt/local/bin:/opt/local/sbin:/usr/bin:/usr/sbin
      ;;
  esac
fi

## Current state of povisioning
CURRENTSTATE=initalize
COMPLATEFILE=/var/svc/zcloud_success
MAILFLAG=true
T_MAILBODY="/tmp/$(basename $0).$$.tmp"
touch $T_MAILBODY


_finalize() {
  EXITCODE=$?
  cat <<EOL >> $T_MAILBODY
Hi,

This is zcloud application automator.

My ipaddress is $IPADDRESS.
Last state is $CURRENTSTATE and exit status is $EXITCODE.

EOL
  echo >> $T_MAILBODY
  echo "========== tail of metadata-execute log" >> $T_MAILBODY
  tail /var/svc/log/smartdc-mdata\:execute.log >> $T_MAILBODY
  if $MAILFLAG
  then
    cat $T_MAILBODY | mailx -s "[$CURRENTSTATE $Z_APP] Zcloud Notification from `hostname`" $MAILTO
  fi
  rm $T_MAILBODY
  exit 0
}

trap _finalize 0

## shared functions
_smf_enabler() {
  if svcs "$1" | grep -q disabled  ; then svcadm enable "$1" ;fi
  if svcs "$1" | grep -q maintenance  ; then svcadm clear "$1" ;fi
}

_get_addr_by_if() {
  ipadm show-addr "$1" -p -o ADDR
}

_mdata_check(){
  if ! mdata-get $1 ; then echo "ERROR_EXIT: missing metadata $1" ; exit 1 ; fi
  export $2="`mdata-get $1`"
}

_splitip() {
  local IFS="/"
  set -- $1
  echo $1
}

# postfix must be running.
_smf_enabler postfix

MAILTO=""
## Set localzone and force reboot
if mdata-get zcloud_notify_to
then
  MAILTO=`mdata-get zcloud_notify_to`
else
  MAILFLAG=false
fi

###
### prepare section
###

CURRENTSTATE=setup_host

IPADDRESS=`_get_addr_by_if net0/_a`

## set hostname. It is no need to reboot.
if mdata-get zcloud_hostname
then
  LNAME=`mdata-get zcloud_hostname`
  if [ ! "`hostname`" == "$LNAME" ] ; then sm-set-hostname ${LNAME} ; fi
else
  ZONENAME=`zonename`
  LNAME=`_splitip ${IPADDRESS//./-}`.${ZONENAME:0:8}.local
  if [ ! "`hostname`" == "$LNAME" ] ; then sm-set-hostname ${LNAME} ; fi
fi

## Set localzone and force reboot
if mdata-get zcloud_timezone
then
  LZONE=`mdata-get zcloud_timezone`
else
  LZONE=Japan
fi
if [ ! "$TZ" == "$LZONE" ] ; then MAILFLAG=false ; sm-set-timezone ${LZONE} && reboot ; fi

###
### main section
###

MDATA_WRAPPER=001
MDATA_USERSCRIPT=/var/svc/mdata-user-script
MDATA_USERDATA=/var/svc/mdata-user-data
CHEF_REPOS=/usr/local/zcloud-application

## add mdata-wapper script to cron
## Notice: this block must keep on top to retry fetch cyclically.


if [ ! -f /opt/local/sbin/mdata_wrapper_${MDATA_WRAPPER}.sh ] ; then
  CURRENTSTATE=setup_wrapper
  cat <<"EOL" > /opt/local/sbin/mdata_wrapper_${MDATA_WRAPPER}.sh
#!/usr/bin/bash
export PATH=/opt/local/bin:/opt/local/sbin:/usr/bin:/usr/sbin
sleep `expr $$ % 60`
svcadm restart mdata:fetch
if svcs mdata:fetch | grep -q maintenance  ; then svcadm clear mdata:fetch ;fi
sleep 2
svcadm restart mdata:execute
if svcs mdata:execute | grep -q maintenance  ; then svcadm clear mdata:execute ;fi
EOL

  chmod u+x /opt/local/sbin/mdata_wrapper_${MDATA_WRAPPER}.sh
  ln -fs /opt/local/sbin/mdata_wrapper_${MDATA_WRAPPER}.sh /opt/local/sbin/mdata_wrapper.sh
fi

if ! grep -q -x "## ZCloud-Application" /var/spool/cron/crontabs/root
then
  CURRENTSTATE=setup_cronjob
  cat << "EOL" >> /var/spool/cron/crontabs/root
## ZCloud-Application
0,10,20,30,40,50 * * * * /opt/local/sbin/mdata_wrapper.sh
EOL
svcadm restart cron
fi

## notifies start setup if new server

if [ ! -f $COMPLATEFILE -a -z $MAILTO ]
then
  cat <<EOL | mailx -s "[Setup started]Zcloud Norify from `hostname`" $MAILTO
Hi,

This is zcloud application automator.

My ipaddress is $IPADDRESS.
A new server has been provisioning. Please wait a few minutes.

EOL
fi

# install joyent_attr_plugin
if [ ! -f /opt/local/etc/ohai/plugins/joyent.rb ] ; then
  CURRENTSTATE=setup_ohai_plugin
  install -d /opt/local/etc/ohai/plugins -m 0755
  curl -skf -o /opt/local/etc/ohai/plugins/joyent.rb https://raw.github.com/ZCloud-Firstserver/ohai_plugin_joyent/master/plugins/joyent.rb
fi

## install chef-solo
if [ ! -f /opt/local/bin/chef-solo ] ; then
  CURRENTSTATE=setup_chef-solo

  pkgin -y install gcc47 scmgit-base gmake ruby193-base ruby193-yajl ruby193-nokogiri ruby193-readline pkg-config

## for smf cookbook
  pkgin -y install libxslt

## install chef
  gem update --system
  gem install --no-ri --no-rdoc bundler
  gem install --no-ri --no-rdoc ohai
  gem install --no-ri --no-rdoc json
  gem install --no-ri --no-rdoc chef
  gem install --no-ri --no-rdoc rb-readline
fi

## get attribute from metadata-api

_mdata_check zcloud_app Z_APP
_mdata_check zcloud_app_repo Z_APP_REPO
if mdata-get zcloud_app_ref ; then _mdata_check zcloud_app_ref Z_APP_REF
else
    Z_APP_REF=master
fi

## exit once
if [ -f $COMPLATEFILE ]
then
  if mdata-get zcloud_app_once
  then
    CURRENTSTATE=running
    exit 0
  fi
fi


## clone or pull application repositoly to local

if [ ! -d ${CHEF_REPOS} ] ; then
  CURRENTSTATE=initalize_git_repository
  git clone ${Z_APP_REPO} ${CHEF_REPOS} -b ${Z_APP_REF}
else
  CURRENTSTATE=update_git_repository
  cd ${CHEF_REPOS}
  ## switch branch by Metadata
  git fetch --all --prune
  git checkout ${Z_APP_REF}
  git pull origin ${Z_APP_REF}
fi

## execute chef-solo
CURRENTSTATE=execute_chef-solo

if chef-solo -j ${MDATA_USERDATA} -c ${CHEF_REPOS}/solo.rb -o "role[${Z_APP}]"
then
  CURRENTSTATE=running
  if [ -f $COMPLATEFILE ]
  then
    MAILFLAG=false
  else
    touch $COMPLATEFILE
  fi
else
  CURRENTSTATE=failure_chef-solo
fi


exit 0
